### ISSUE TITLE
[BUG] Conflict with `CLIP NegPip` from `ComfyUI-ppm` due to ComfyUI Core Update (#10362)

### ISSUE THREAD

**Issue by higana-yogana** [asagi4/comfyui-prompt-control#135]
**Describe the bug**
I am reporting a compatibility issue between `comfyui-prompt-control` and the `CLIP NegPip` node from `ComfyUI-ppm`.

In the past, these two extensions were designed to work together through mutual effort by the authors. However, this established cooperation seems to have broken recently due to changes in the ComfyUI core.

Specifically, the issue began after the core update:
**"Fix order of inputs nested merge_nested_dicts (#10362)" (Commit: `493b81e`)**

In this commit, the order of arguments in the recursive call within `merge_nested_dicts` in `comfy/patcher_extension.py` was swapped:

```python
-            merged_dict[key] = merge_nested_dicts(value, curr_value)
+            merged_dict[key] = merge_nested_dicts(curr_value, value)
```

This change in how nested dictionaries (like those used for patches and hooks) are merged appears to cause `COUPLE` to malfunction when `NegPip` is active in the same workflow.

**Additional Context**
- It is possible that fixes are required on the `ComfyUI-ppm` side as well.
- Since this can be bypassed by simply not using `NegPip` when `COUPLE` is needed, it is not an urgent issue. I am reporting this primarily to keep you informed of the current situation.

I rely heavily on `comfyui-prompt-control` and it has been an immense help to my workflow. I deeply appreciate the effort that originally went into making these nodes compatible, and I sincerely hope that this harmony can be restored in the future.

*Note: This technical analysis was generated with the assistance of Gemini (AI). Please use it for reference purposes.*

**Comment by asagi4** [comment]
I can't find any calls to merge_nested_dicts in either my code or the ppm nodepack, so it seems likely that it's used indirectly by some other ComfyUI function instead.

Perhaps Comfy-Org/Comfyui@493b81e breaks the usage of that function somewhere else in ComfyUI.

**Comment by higana-yogana** [comment]
As you can see, this is not a fundamental solution but rather a workaround.
Since using NegPip and COUPLE together is a somewhat rare use case, it might be sufficient to just add a note to the documentation and leave the behavior as is.

**Comment by higana-yogana** [comment]
Yes, you are absolutely right.
Originally, I tried to solve it using the existing flags/mechanisms, but it was beyond my capability.
So I added a flag that I could easily track and forced a different route just to verify the behavior.
This confirmed that if we can just overcome that specific barrier (the patch conflict/ordering), the rest of the logic
works correctly.

> I wonder if previously the attention functions were called in the order [negpip, pc_couple] and now it's [pc_couple,
negpip].

I agree that this reversal in priority is likely the core issue.
It seems pc_couple's processing is being overwritten or effectively cancelled out by negpip coming later (or vice
versa in a way that breaks the stack).
And I suspect that the change in argument order in merge_nested_dicts is what caused this strange reversal in
how the patch lists are merged.

When I reached this point in my investigation, I found myself wondering, "Where should I even report this? Is my
understanding even correct? And is this issue worth the cost to fix?" which led me to put it aside until now.
I apologize for the confusion.

**Comment by Booyaka101** [comment]
Picking up where this left off — wanted to contribute something useful since you mentioned no time. I read through `attention_couple_ppm.py` to check if your "patch order reversed" hypothesis lines up with what ComfyUI core's `merge_nested_dicts` change actually does, and I think it does, mechanically:

**Pre-`493b81e`** — `merge_nested_dicts(value, curr_value)` — for list-valued keys like `attn2_patch`, the *new* hook's list was extended onto the *existing* list. So when AttentionCouple registered after NegPip, `attn2_patch` ended up as `[negpip_patch, couple_patch]` and ran in that order.

**Post-`493b81e`** — `merge_nested_dicts(curr_value, value)` — args swapped, so the *existing* list is extended onto the *new* one. `attn2_patch` becomes `[couple_patch, negpip_patch]`, and couple's masked-attention math now operates on un-negpip'd V before NegPip gets a chance to swap halves. That matches the symptom (regions ignored, image looks like couple did nothing) — couple's per-region attention happens, but downstream negpip then re-modifies V across the whole batch and the per-region weighting gets washed out.

So I think you're right that this is fundamentally a ComfyUI-core ordering regression, but I don't think a fix needs to land in core — it can live in `AttentionCoupleHook` by not relying on dict-merge order at all. Two sketches I considered:

**(a) Post-merge resort in `on_apply_hooks`.** After `super().on_apply_hooks(model, transformer_options)`, walk `transformer_options["patches"]["attn2_patch"]`, find any patch that isn't ours, and move it to the front. Robust to any future merge-order change in core but a bit invasive (mutating a list ComfyUI just composed for us).

**(b) Compose explicitly at hook-application time.** Skip registering `attn2_patch` / `attn2_output_patch` via the `transformers_dict` mechanism entirely. Instead, in `on_apply_hooks` read whatever `attn2_patch` already exists in `transformer_options["patches"]`, build a new list that is `existing + [Proxy(self.attn2_patch)]`, and overwrite the entry. Same for `attn2_output_patch`. This sidesteps `merge_nested_dicts` for these specific keys — couple's patches are guaranteed to run after whatever was there.

Both should restore `[negpip, couple]` order regardless of which way Comfy core decides to merge. (a) is one-extra-pass, (b) is the structural fix. I'd lean (b) because it removes the implicit dependency on dict-merge semantics — that ordering was always a coincidence of `merge_nested_dicts`'s behavior, never something the hook system documented.

I don't have a NegPip+COUPLE workflow on hand to verify either patch end-to-end, and I don't want to ship a fix to your repo blind given the symptom is subtle. Two ways I can help if useful:

- **Draft PR for review:** I'll put up (b) as a draft PR with the diff, and you (or @higana-yogana, who has the failing workflow) gets the final say on testing. Marked draft so there's no merge pressure.
- **Just leave the analysis here:** if you'd rather take this and write it your own way, all good — happy to be a sideline contributor. You know this code far better than I do.

Either way, hopefully this saves a debug round.

**Comment by asagi4** [comment]
Feel free to send in patches. I haven't had much incentive to look into this since it's not a bug I'm really hitting myself.

**Comment by asagi4** [comment]
This patch to ComfyUI makes things work (no changes needed in prompt control)
```
From 35c83ef28d007af0f87e54ca7876ccecc9af6a11 Mon Sep 17 00:00:00 2001
From: asagi4 <130366179+asagi4@users.noreply.github.com>
Date: Fri, 1 May 2026 21:55:48 +0300
Subject: [PATCH] Apply hook patches on top of model patches

---
 comfy/samplers.py | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/comfy/samplers.py b/comfy/samplers.py
index 0a4d062d..a4a5671d 100755
--- a/comfy/samplers.py
+++ b/comfy/samplers.py
@@ -300,9 +300,8 @@ def _calc_cond_batch(model: BaseModel, conds: list[list[dict]], x_in: torch.Tens

             transformer_options = model.current_patcher.apply_hooks(hooks=hooks)
             if 'transformer_options' in model_options:
-                transformer_options = comfy.patcher_extension.merge_nested_dicts(transformer_options,
-                                                                                 model_options['transformer_options'],
-                                                                                 copy_dict1=False)
+                # Transformer options from hooks should stack *after* any patches on the model itself.
+                transformer_options = comfy.patcher_extension.merge_nested_dicts(model_options["transformer_options"], transformer_options, copy_dict1=True)

             if patches is not None:
                 transformer_options["patches"] = comfy.patcher_extension.merge_nested_dicts(
--
2.47.3
```
I'm not sure if that has any other effects, but I think it makes sense for hooks to apply on top of what's on the model instead of the other way around.

**Comment by Booyaka101** [comment]
Nice — that's exactly the swap I was hand-waving toward in the close on #145, glad you found it works empirically. Two minor procedural questions before this lands somewhere users can pick it up:

1. Are you planning to file this against `comfyanonymous/ComfyUI` upstream, or would a PR there be welcome? Happy to file it with proper credit (commit by you preserved, PR description quoting your patch and the architectural trace from the thread here / from my close on #145), if you'd rather not deal with the core review cycle. Totally fine if you'd rather file it yourself — just don't want to step on it.

2. The companion swap at `samplers.py:308` (the patches-specific `merge_nested_dicts` call right below the one your patch fixes) has the same dict1/dict2 polarity issue if I'm reading it right. Did you test against that or leave it as-is? Looking at it now:

   ```py
   if patches is not None:
       transformer_options["patches"] = comfy.patcher_extension.merge_nested_dicts(
           transformer_options.get("patches", {}),  # dict1: now "model + hooks"
           patches                                   # dict2: per-cond patches
       )
   ```

   With your line-303 swap landed, `transformer_options` at this point already contains `model_options["transformer_options"]["patches"]["attn2_patch"] = [negpip]` plus the hook's `[couple]`. The `patches` arg here is per-cond control patches, not NegPip-related — so possibly fine to leave alone. But worth a quick eyeball before the upstream PR.

Let me know your preference and I'll either file it or stand down.

**Comment by asagi4** [comment]
I think the line 308 patches are applied in the correct order. The final patch order will be model + hooks + cond patches, which seems correct to me.

Feel free to create a PR yourself, though a more proper fix might be to pass a clone of `model_options['transformer_options']` to the hook (via the existing transformer_options parameter to on_apply_hooks) so that the hook has control over the order in which it inserts its patches. The default TransformerOptionsHook behaviour will essentially then do the model + hook patch merge and the if block at line 302 can be removed.

**Comment by asagi4** [comment]
apply_hooks already takes a transformer_options parameter apparently, so this seems like it works too, but again, not very well tested for other effects:
```
diff --git a/comfy/samplers.py b/comfy/samplers.py
index 0a4d062d..c49b3518 100755
--- a/comfy/samplers.py
+++ b/comfy/samplers.py
@@ -298,11 +298,7 @@ def _calc_cond_batch(model: BaseModel, conds: list[list[dict]], x_in: torch.Tens
             c = cond_cat(c)
             timestep_ = torch.cat([timestep] * batch_chunks)

-            transformer_options = model.current_patcher.apply_hooks(hooks=hooks)
-            if 'transformer_options' in model_options:
-                transformer_options = comfy.patcher_extension.merge_nested_dicts(transformer_options,
-                                                                                 model_options['transformer_options'],
-                                                                                 copy_dict1=False)
+            transformer_options = model.current_patcher.apply_hooks(hooks=hooks, transformer_options=model_options.get("transformer_options", {}).copy())

             if patches is not None:
                 transformer_options["patches"] = comfy.patcher_extension.merge_nested_dicts(
```

**Comment by Booyaka101** [comment]
Filed upstream as draft: [comfyanonymous/ComfyUI#13671](https://github.com/Comfy-Org/ComfyUI/pull/13671). Went with the smaller arg-swap (Patch A) for the smallest possible review surface, with your kwarg-threading variant (Patch B) mentioned in the body as the architectural alternative if reviewers prefer it. Commit authored by you, co-authored by me. Marked draft so there's no review pressure on either of you.

I included a self-contained verification script in the PR body that imports the real `merge_nested_dicts` and confirms pre-fix produces `[couple, negpip]` and the patch produces `[negpip, couple]` — proves the merge order at the dict layer without needing to run diffusion.

Will track the upstream review and ping you here if reviewers want to discuss anything that needs your call. Thanks for the patches and the diagnostic work.
